pkg/controller/controller_instance.go

```diff
@@ -1833,45 +1196,213 @@ func shouldStartOrphanMitigation(statusCode int) bool {
 
 // serviceInstanceHasExistingBindings returns true if there are any existing
 // bindings associated with the given ServiceInstance.
-func (c *controller) serviceInstanceHasExistingBindings(instance *v1beta1.ServiceInstance) (bool, error) {
+func (c *controller) checkServiceInstanceHasExistingBindings(instance *v1beta1.ServiceInstance) error {
 	bindingLister := c.bindingLister.ServiceBindings(instance.Namespace)
 
 	selector := labels.NewSelector()
 	bindingList, err := bindingLister.List(selector)
 	if err != nil {
-		return false, err
+		return err
 	}
 	for _, binding := range bindingList {
 		if instance.Name == binding.Spec.ServiceInstanceRef.Name {
-			return true, nil
+			return &operationError{
+				reason:  errorDeprovisionBlockedByCredentialsReason,
+				message: "All associated ServiceBindings must be removed before this ServiceInstance can be deleted",
+			}
 		}
 	}
 
-	return false, nil
+	return nil
 }
 
-// cloneServiceInstance clones the given ServiceInstance. This is so that the
-// reconciler may modify the cloned resource without accidentally mutating the
-// cache.
-func cloneServiceInstance(instance *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, error) {
-	clone, err := api.Scheme.DeepCopy(instance)
+// requestHelper is a helper struct with properties common to multiple request
+// types.
+type requestHelper struct {
+	ns                   *corev1.Namespace
+	parameters           map[string]interface{}
+	inProgressProperties *v1beta1.ServiceInstancePropertiesState
+	originatingIdentity  *osb.OriginatingIdentity
+	requestContext       map[string]interface{}
+}
+
+// prepareRequestHelper is a helper function that generates a struct with
+// properties common to multiple request types.
+func (c *controller) prepareRequestHelper(instance *v1beta1.ServiceInstance, serviceClass *v1beta1.ClusterServiceClass, servicePlan *v1beta1.ClusterServicePlan) (*requestHelper, error) {
+	rh := &requestHelper{}
+
+	if utilfeature.DefaultFeatureGate.Enabled(scfeatures.OriginatingIdentity) {
+		originatingIdentity, err := buildOriginatingIdentity(instance.Spec.UserInfo)
+		if err != nil {
+			return nil, &operationError{
+				reason:  errorWithOriginatingIdentity,
+				message: fmt.Sprintf("Error building originating identity headers: %v", err),
+			}
+		}
+		rh.originatingIdentity = originatingIdentity
+	}
+
+	if instance.Status.AsyncOpInProgress || instance.Status.OrphanMitigationInProgress || instance.DeletionTimestamp != nil {
```

Comment:
I would like to see a function that takes the instance as a parameter and returns the transition that the controller is performing on the instance--i.e., Add, Update, Delete, or Poll. This has a few benefits that I can see.

1) It puts in one place the logic for determining which transition to perform. Currently, the logic is in at least two places. It is in the `reconcileServiceInstance` function and here.
2) It makes it clearer to understand what is happening here without having to remember what the conditionals represent here.
3) It would assuage @n3wscott on his concern about the default branch in `reconcileServiceInstance` since the switch would then becoming a switch on the transition type enumerations, making a default that does nothing more appropriate.

Reply:
Sure, let me take a stab at that.